threefoldtecharchive/jumpscaleX_core
# Development secrethandshake

The main change introduce in this PR is the fact that to be able to communicate with a threebot server, the client needs to know the public key of the 3bot he is trying to join. This is done by looking up the public key in the phonebook.

Then during initialization of the connection, keys are exchanged to authenticate both parties and a session key is generated to encrypt the communication so only both peer can read the data flowing through.

There is one thing to take in account though. Since the phonebook is also reached over gedis, you at least needs to know the public key of the explorers hosting the phonebook.
In this PR I've hardcoded the public key of both testnet and production net, but we might want to publish this also somewhere else. Think of this as some sort of bootstrap you need to be aware of before being able fully communicate with any 3bot

How to test this:
1. prepare 3 jsx container, 1 will be the explorer, 2 other will be users 3bots
2. On the explore container, read the public key: 
```python
j.data.nacl.default.verify_key_hex
```
3. On the explorer starts the 3bot server: 
```python
kosmos -p "j.servers.threebot.local_start_explorer(background=True)"
```
4. on the 2 users jsx container, configure the explorer address and public key. 
```python
j.clients.threebot.explorer_addr_set('172.17.0.2','34c77fdf6c5ef24d5a6981be06f9109ba83b7e306cfa1d8141ce5f572b647cbeb') (adapt the values here...)
```
5. on the 2 users jsx container, register their identity. 
```python
j.tools.threebot.init_my_threebot(name='user1',email='user1@mail.com',description='')
j.tools.threebot.init_my_threebot(name='user2',email='user1@mail.com',description='')
```
6. start bot server of both users jsx container:  
```python
j.servers.threebot.local_start_3bot()
```
7. try to connect from user1 to user2: 
```python
cl = j.clients.threebot.client_get(threebot='user2')
cl.actors_all
```
8. You can use wireguard to check the traffic between the 2 user container is encrypted
